### ISSUE TITLE
Problems with reusing finder.findPath()?

### ISSUE THREAD

**Issue by samlittler** [qiao/PathFinding.js#179]
My usage: When findPath returns an empty array because the destination is impassable, I'm testing various surrounding nodes to see if I can reach them instead.

Using AStar algorithm, the first time I call this function I receive a list of nodes as expected.

On subsequent calls using the same instance of AStarFinder with the same grid but different coordinates, I seem to receive empty arrays in response.

**Comment by samlittler** [comment]
Just an update on what I've discovered here...

It seems as though `AStarFinder.findPath()` manipulates the `PFGrid` in some way.

If instead I do this:

`finder.findPath(this.x, this.y, x, y, grid.clone());`

I get the expected behaviour.

**Comment by brean** [comment]
Yes, the implementation in Pathfinding.js stores which nodes it has seen, so the grid needs to be cloned before or reset afterwards (see the `opened` and `closed` variables in the finder).
In theory this could be changed, the [pseudo code at Wikipedia](https://en.wikipedia.org/wiki/A*_search_algorithm#Pseudocode) shows a version that stores everything in an own objects/arrays without modifying the grid. 
I think both versions have their benefits and drawbacks when it comes to execution time and memory usage. I think it should not make much of a difference, but it maybe makes sense to implement the pseudo code and run some benchmarks.
